#### Integrative analysis identifies potential DNA methylation biomarkers for pan-cancer diagnosis and prognosis

##### RNAseq数据中等于0的值替换为最小正值

```R
idmapping <- read.table(file = "fanai/jiajihua/COAD_RNAseq", 
                        sep = "\t", header = T, stringsAsFactors = F)
rownames(idmapping) <- idmapping$sample
idmapping <- idmapping[,-1]
minelement <- c()
for(i in colnames(idmapping)){
  tmp <- sort(idmapping[,i])
  minelement <- c(minelement,tmp[tmp>0][1])
  
}
minelement <- sort(minelement)[1]
coad <- idmapping
coad[coad==0] <- minelement

idmapping <- read.table(file = "fanai/jiajihua/ESCA_RNAseq", 
                        sep = "\t", header = T, stringsAsFactors = F)
rownames(idmapping) <- idmapping$sample
idmapping <- idmapping[,-1]
minelement <- c()
for(i in colnames(idmapping)){
  tmp <- sort(idmapping[,i])
  minelement <- c(minelement,tmp[tmp>0][1])
  
}
minelement <- sort(minelement)[1]
esca <- idmapping
esca[esca==0] <- minelement

idmapping <- read.table(file = "fanai/jiajihua/STAD_RNAseq", 
                        sep = "\t", header = T, stringsAsFactors = F)
rownames(idmapping) <- idmapping$sample
idmapping <- idmapping[,-1]
minelement <- c()
for(i in colnames(idmapping)){
  tmp <- sort(idmapping[,i])
  minelement <- c(minelement,tmp[tmp>0][1])
  
}
minelement <- sort(minelement)[1]
stad <- idmapping
stad[stad==0] <- minelement

idmapping <- read.table(file = "fanai/jiajihua/READ_RNAseq", 
                        sep = "\t", header = T, stringsAsFactors = F)
rownames(idmapping) <- idmapping$sample
idmapping <- idmapping[,-1]
minelement <- c()
for(i in colnames(idmapping)){
  tmp <- sort(idmapping[,i])
  minelement <- c(minelement,tmp[tmp>0][1])
  
}
minelement <- sort(minelement)[1]
read <- idmapping
read[read==0] <- minelement
```

#### 去除性染色体上的探针

```R
idmapping <- read.table(file = "COAD_methy450", sep = "\t", header = T, stringsAsFactors = F)
rownames(idmapping) <- idmapping$sample
idmapping <- idmapping[,-1]
zhuanhuan <- read.table(file = "illuminaMethyl450_hg19_GPL16304_TCGAlegacy", sep = "\t", header = F, stringsAsFactors = F)
colnames(zhuanhuan) <- c('sample','gene','chrom','chromStart','chromEnd','strand')

#不在性染色体上的探针
zhuanhuan1 <- zhuanhuan[zhuanhuan$chrom != 'chrX',]
zhuanhuan2 <- zhuanhuan1[zhuanhuan1$chrom != 'chrY',]
zhuanhuan <- zhuanhuan2
idmapping <- idmapping[zhuanhuan$sample,]
```

#### 去除NA数量大于0.3的样本

```R
total<- dim(idmapping)[1]
total <- as.integer(total)
col <- c()
for(i in colnames(idmapping)){
  naCount <- table(is.na(idmapping[,i]))['TRUE']
  naCount <- as.integer(naCount)
  if(is.na(naCount)){
    naCount <- as.integer(0)
  }
  ratio <- naCount/total
  if(ratio>0.3)
    col <- c(col,i)
}
idmapping <- idmapping[,!colnames(idmapping) %in% col]
```

#### 去除NA数量大于0.3的探针

```R
samples <- dim(idmapping)[2]
samples <- as.integer(samples)
roW <- c()
count <- 0
for(i in rownames(idmapping)){
  naCount <- table(is.na(idmapping[i,]))['TRUE']
  naCount <- as.integer(naCount)
  if(is.na(naCount)){
    naCount <- as.integer(0)
  }
  ratio <- naCount/samples
  if(ratio>0.3)
    roW <- c(roW,i)
  count <- count+1
  print(count)
}
idmapping <- idmapping[!rownames(idmapping) %in% roW,]
```

#### 甲基化差异分析

```R
library(impute)
library(ChAMP)
tum <- c()
nor <- c()
for(i in colnames(expr)){
    tmp <- as.numeric(substr(i,14,15))
    tmp1 <- substr(i,1,12)
    if(tmp < 10){
    	tum <- c(tum,tmp1)
    }
    else{
    	nor <- c(nor,tmp1)
    }
}
jiao <- intersect(tum,nor)
jiaot <- c()
for(i in jiao){
    jiaot <- c(jiaot,paste(i,'.01',sep=''))
}
jiaon <- c()
for(i in jiao){
    jiaon <- c(jiaon,paste(i,'.11',sep=''))
}
tumor <- expr[,jiaot]
normal <- expr[,jiaon]
expr <- cbind(tumor,normal)
group_list <- ifelse(as.numeric(substr(colnames(expr),14,15))<10,'tumor','normal')
info <- data.frame(Sample=colnames(expr),Group=group_list)
rownames(info) <- info[,1]
expr <- as.matrix(expr)
expr <- impute.knn(expr)
expr <- expr$data
expr <- expr+0.00001
identical(colnames(expr),rownames(info))
myLoad <- champ.filter(beta=expr,pd=info)
myNorm <- champ.norm(beta=myLoad$beta,arraytype='450k',cores=5)
pD=myLoad$pd
beta.m=myNorm
group_list=myLoad$pd$Group
myDMP <- champ.DMP(beta = myNorm,pheno=group_list)
deg=myDMP[[1]]
deg$g=ifelse(abs(deg$logFC) < 0.2,'stable',ifelse(deg$logFC > 0.2,'UP','DOWN'))
deg$rowname <- rownames(deg)
library(openxlsx)
write.xlsx(deg,file='coad_chayi.xlsx')
```

